### DISCUSSION TITLE
Microsoft Azure Identity provider SSO Integration with Directus not working

### DISCUSSION THREAD

**Discussion by AmirEzati** [directus/directus#13137] [Archive]
I have tried to employ Microsoft AD as identity provider for Directus but have failed to do so.
The authentication with my microsoft account is done successfully each time, but Directus never allows the account to get into the service.
Different versions of Directus package were also installed but no luck yet.

### To Reproduce

- Install Directus via npm and open the project with a code editor.

- place the  below key/value pairs in the .env file

AUTH_PROVIDERS="microsoft"
PUBLIC_URL= "http://localhost:8055"
AUTH_MICROSOFT_DRIVER = "openid"
AUTH_MICROSOFT_CLIENT_ID = "from Azure"
AUTH_MICROSOFT_CLIENT_SECRET = "from Azure"
AUTH_MICROSOFT_ISSUER_URL = "https://login.microsoftonline.com/{tenantId}/v2.0/.well-known/openid-configuration"
AUTH_MICROSOFT_ALLOW_PUBLIC_REGISTRATION="true" 
AUTH_MICROSOFT_DEFAULT_ROLE_ID="{an_available_role_id}"
AUTH_MICROSOFT_ICON="microsoft"
AUTH_MICROSOFT_IDENTIFIER_KEY = "sub" // tried with email or upn as well
AUTH_MICROSOFT_SCOPE="openid profile email"

### Errors Shown

over the address bar I get: **http://localhost:8055/admin/login?reason=SERVICE_UNAVAILABLE**
in terminal I get the below error messages:
 **- OPError [OpenID] Unknown OP error
 - Service returned unexpected response
 - Auth provider not configured
-  request errored POST 503 /auth/refresh** 

and when I try to call the callback url directly on the addressbar to see some output:
in the address bar: http://localhost:8055/auth/login/microsoft/callback?code={the code from azure}
the error shown: 
**{"errors":[{"message":"Invalid user credentials.","extensions":{"code":"INVALID_CREDENTIALS"}}]}**

<details>
<summary>More info</summary>

### What version of Directus are you using?

9.10.0

### What version of Node.js are you using?

16.14.0

### What database are you using?

Postgres 14

### What browser are you using?

Chrome

### How are you deploying Directus?

running locally

</details>

**Comment by azrikahar** [comment]
> Auth provider not configured

This error seems odd, but can you try removing the spaces before and after the equal signs in your `.env` file? Example:

```diff
- AUTH_MICROSOFT_DRIVER = "openid"
+ AUTH_MICROSOFT_DRIVER="openid"
```

unless of course if the spaces only exists _after_ you paste them here (just a formatting issue), feel free to ignore this.

You can also try referring to this config shared by another community memter as another form of sanity check: https://github.com/directus/directus/discussions/10511#discussioncomment-1814764

**Reply by AmirEzati** [reply]
yes, spaces between key and values are removed.
I ran in cognito as well but the same error

**Reply by azrikahar** [reply]
> I ran in cognito as well but the same error

Does "Auth provider not configured" still show up as the error?

---

@aidenfoxx thoughts on what could be the issue here?

**Reply by AmirEzati** [reply]
it is configured and the button to authenticate with microsoft is appearing. 
Then I click to authenticate with microsoft, I successfully get authenticated but on the way back to Directus, I get this error

**Reply by aidenfoxx** [reply]
"Unknown OP error" is a server error. Usually that the server provided an unexpected response. You can see more log details by adding the following config options:

```
LOG_LEVEL="trace"
LOG_STYLE="raw"
```

**Reply by aidenfoxx** [reply]
Also, is your `PUBLIC_URL` set to a fully qualified domain name? That's veeeeeeery important

**Reply by AmirEzati** [reply]
Yes it is already set to http://localhost:8055

**Reply by aidenfoxx** [reply]
@AmirEzati If you manage to get things fixed please drop a message explaining how for others in the future. 😄

**Reply by AmirEzati** [reply]
Sure, I will do. 
perhaps it is worth mentioning that SSO with Auth0 works fine but the issue with Microsoft as an Auth provider remains.

**Comment by AmirEzati** [comment]
Does anybody have any suggestion?
Have you ever integrated microsoft Azure as the identity provider? if yes, can we compare the configurations?
The integration should be simple and straightforward.
any ideas?

**Reply by aidenfoxx** [reply]
Just verified with [the following configuration](https://docs.directus.io/configuration/sso-examples/#microsoft-azure) and everything works correctly:

```
AUTH_MICROSOFT_DRIVER="openid"
AUTH_MICROSOFT_CLIENT_ID="<client_id>"
AUTH_MICROSOFT_CLIENT_SECRET="<client_secret>"
AUTH_MICROSOFT_ISSUER_URL="https://login.microsoftonline.com/<my_tennant_id>/v2.0/.well-known/openid-configuration"
AUTH_MICROSOFT_IDENTIFIER_KEY="email"
```

**Reply by AmirEzati** [reply]
no matter locally, or via helm charts, the result is always **http://localhost:8055/admin/login?reason=SERVICE_UNAVAILABLE**

my exact config:

  - name: AUTH_PROVIDERS
    value: "microsoft"
  - name: AUTH_MICROSOFT_DRIVER
    value: "openid"
  - name: AUTH_MICROSOFT_CLIENT_ID
    value: "<client_id>"
  - name: AUTH_MICROSOFT_CLIENT_SECRET
    value: "<client_secret>"
  - name: AUTH_MICROSOFT_ISSUER_URL
    value: "https://login.microsoftonline.com/<my_tennant_id>/v2.0/.well-known/openid-configuration"
  - name: AUTH_MICROSOFT_IDENTIFIER_KEY
    value: "email"

**Reply by aidenfoxx** [reply]
Are you using the correct client ID? You need your Tennant client ID. Not the ID that comes with the secret.

And obviously you need your tennant ID in the "ISSUER_URL". Also make sure you have a valid redirect URL configured in azure. It should be "http://<your_directus_url>/auth/login/microsoft/callback"

**Reply by AmirEzati** [reply]
yes, the clinentId is the id of the registered app in the azure portal,
the tenantId and the callback url also follow the exact recommended ones:

http://localhost:8055/auth/login/microsoft/callback
![Bildschirmfoto 2022-05-19 um 17 06 55](https://user-images.githubusercontent.com/30486048/169329119-4bf66c2d-aad1-4a81-90e5-20f2432655fd.png)

**Reply by aidenfoxx** [reply]
Well then I don't know what to tell you. Sounds to me like it's either a configuration issue somewhere, or your internal network is blocking Directus from communicating with Azure. If your network requires a proxy that could be what's causing the issue.

**Comment by AmirEzati** [comment]
the below screenshots are two of the errors that we caught in the logs of the directus app: could you please help us with them?
it would great if you share the manifest of your sample registered microsoft app in Azure, so we can compare it with our own one and get closer to the solution.
Is there any detailed documentation on SSO with Azure that goes through all the steps required to integrate microsoft azure as an identity provider?

<img width="892" alt="Bildschirmfoto 2022-05-25 um 15 42 05" src="https://user-images.githubusercontent.com/30486048/170276528-7d829f82-2967-4270-8d64-a37158653cee.png">

<img width="1029" alt="Bildschirmfoto 2022-05-25 um 15 43 21" src="https://user-images.githubusercontent.com/30486048/170276809-e7d39159-63a4-430e-bd68-06b201e9a8f7.png">

**Reply by rijkvanzanten** [reply]
I'm not too familiar with the intricacies of Azure Identity OpenID, but based on the shared error message, have you tried omitting the client ID and client secret environment variables?

**Reply by aidenfoxx** [reply]
Disable "Treat application as public client" or "Allow public client flows" in your Azure app "Authentication" settings.

**Reply by aidenfoxx** [reply]
This is the configuration that's working for me. Basically the default:

![azure](https://user-images.githubusercontent.com/5548518/170280438-6d983207-6fb3-4933-a8e3-94ec713206fa.png)

I also have a single client/secret configured for the app.

**Reply by aidenfoxx** [reply]
The issue is almost certainly a configuration issue on the Azure side. We use a standard OpenID "code" authorization flow. The error indicates your Azure configuration is not compatible with that.

**Reply by AmirEzati** [reply]
> This is the configuration that's working for me. Basically the default:
> 
> ![azure](https://user-images.githubusercontent.com/5548518/170280438-6d983207-6fb3-4933-a8e3-94ec713206fa.png)
> 
> I also have a single client/secret configured for the app.

thanks for sharing this.
I tried with the default ones but still Service_Unavailabe

**Reply by AmirEzati** [reply]
> The issue is almost certainly a configuration issue on the Azure side. We use a standard OpenID "code" authorization flow. The error indicates your Azure configuration is not compatible with that.

thanks alot for the info.
 is there any documentation on the steps to adjust the registered app for the directus?

**Reply by aidenfoxx** [reply]
Unfortunately at this point it's hard for us to say what's going on. We use the "code" flow, as I mentioned, which requires the client ID and secret. The first error you showed indicates that the azure configuration didn't support that. Without seeing any other errors you are getting it's impossible to speculate.

The majority of the OpenID flow is handled by the `openid-client` library we use, which is where the errors is triggered. If you want a better understanding about what's happening you might get better answers there.

**Reply by aidenfoxx** [reply]
@AmirEzati I'm no Azure expert. We don't provider extensive configuration documentation for individual providers. We only provide a basic list of SSO configurations that have been tested working with provider default settings.

**Comment by AmirEzati** [comment]
yes, then we get invalid token error
<img width="654" alt="Bildschirmfoto 2022-05-25 um 15 50 47" src="https://user-images.githubusercontent.com/30486048/170278418-9425e295-4606-4ac9-a8aa-b0ff436c8530.png">

**Comment by AmirEzati** [comment]
@aidenfoxx 
this is the latest azure config, we have a valid clientId+Secret
![Bildschirmfoto 2022-05-25 um 16 23 13](https://user-images.githubusercontent.com/30486048/170285387-4b7a2d9a-0e02-4c29-8c74-e20598d66fd3.png)

and this is the env variables we are providing through directus service:

AUTH_PROVIDERS="microsoft"

PUBLIC_URL="http://localhost:8055"

AUTH_MICROSOFT_DRIVER="openid"

AUTH_MICROSOFT_CLIENT_ID="..."

AUTH_MICROSOFT_CLIENT_SECRET="...."

AUTH_MICROSOFT_ISSUER_URL="https://login.microsoftonline.com/our_tenant_id/v2.0/.well-known/openid-configuration"

AUTH_MICROSOFT_ALLOW_PUBLIC_REGISTRATION="true" 

AUTH_MICROSOFT_ICON="microsoft"

AUTH_MICROSOFT_IDENTIFIER_KEY="email"

we have added and removed  some other env variables but lets say with this last ones, we get "Invalid Token" error.
it is worth mentioning that the user is created in the directus database, but still we canot login with the sso user.

**Reply by aidenfoxx** [reply]
Does your user have any existing "auth_data" in the database? I would recommend deleting the user and trying again.

**Reply by AmirEzati** [reply]
@aidenfoxx  yes, tried once more by deleting the created user but the same invalid token error
could you share the manifest of your registered app?

**Reply by aidenfoxx** [reply]
Sure thing. https://foxx.io/azure-manifest.json

**Reply by aidenfoxx** [reply]
Here is also my Azure OpenID `.well-known` configuration if it helps. https://foxx.io/azure-tenant.json

**Reply by AmirEzati** [reply]
@aidenfoxx  thanks a lot, the issue is resolved.
I figured out that each time the sso login flow fails, a new user is created in the database and in order to make the further configurations effective, we needed actually to remove the user each time manually from the database and then test the new configs. finally after a lot of tries,  it worked perfectly.
in summary, the environment variables are exactly the same ones that I posted previously, and the manifest is also entirely similar to  the one you provided.
so All Good.
Thanks everybody who tried to help.
enjoy your time.

**Reply by aidenfoxx** [reply]
Makes sense. When the user is created we take the "refresh_token" we get from the provider and store it against the user. I guess when you change the configuration, the "refresh_token" is no longer valid.

**Reply by vantao17361** [reply]
Hi @AmirEzati 
Current I have the issue same your case could you share to me detail the environment variables in Directus and Entra ID, I also use Open ID authen, and getting the issue 
![image](https://github.com/user-attachments/assets/68abf47b-770d-4dc7-9943-dc32349478a9)
 Many thanks
Parker

**Comment by AmirEzati** [comment]
Hello Parker,

Back then I had followed the steps listed in the guidlines of directus.
My problem were solved only after I removed the user from the user table
and let the user be freshly created after any configuration changes.
Hope this helps.

Regards,
Amir

**Reply by vantao17361** [reply]
> Hello Parker, Back then I had followed the steps listed in the guidlines of directus. My problem were solved only after I removed the user from the user table and let the user be freshly created after any configuration changes. Hope this helps. Regards, Amir
> […](#)
> On Fri, Mar 21, 2025, 05:34 vantao17361 ***@***.***> wrote: Hi @AmirEzati <https://github.com/AmirEzati> Current I have the issue same your case could you share to me detail the environment variables in Directus and Entra ID, I also use Open ID authen, and getting the issue image.png (view on web) <https://github.com/user-attachments/assets/68abf47b-770d-4dc7-9943-dc32349478a9> Many thanks Parker — Reply to this email directly, view it on GitHub <[#13137 (reply in thread)](https://github.com/directus/directus/discussions/13137#discussioncomment-12571680)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHIS4IEEH3EO3YH2VBKJFE32VNXR3AVCNFSM6AAAAABZO4HAYCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENJXGE3DQMA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Hi @AmirEzati 
Thanks for your reply. I also deleted the users created automatically in the Directus Admin panel, but after I tried, the issue was present. May I know the version of Directus you are using?

Regards,
Parker

**Comment by vantao17361** [comment]
Hi @AmirEzati 
My issue has been resolved. The cause of error "INVALID_CREDENTIALS" is the user was not set "active" in Directus. I have manual went to Directus Admin and navigate that user and then set it active as screeshot
![image](https://github.com/user-attachments/assets/4474b6e7-d8f3-40af-a2a1-8cb8161d6c23)
I'm wondering whether have a method that we can set it in the Directus .env file
